File: 78.js
#!/usr/bin/env node

// Coin partitions

var program = require('commander');

program
    .version('0.1.0')
    .description('Coin partitions')
    .option('-d, --divisor <int>', 'The target divisor', Number, 1000000)
    .parse(process.argv);

var pentagonal = (function() {
    var memo = {};

    return function(n) {
        if (!(n in memo)) {
            memo[n] = (3 * n * n - n) / 2;
        }
        return memo[n];
    };
})();

var partitions = (function() {
    var memo = {
        '0': 1,
    };

    return function(n) {
        if (n in memo) {
            return memo[n];
        }

        var parts = 0;
        var k = 1;
        while (true) {
            var pent = pentagonal(k);
            if (n - pent < 0) {
                break;
            }
            parts += Math.pow(-1, k - 1) * partitions(n - pent);
            parts = parts % program.divisor;
            k = k > 0 ? -k : -k + 1;
        }

        memo[n] = parts;
        return parts;
    };
})();

var n = 1;
while (partitions(n)) {
    n++;
}

console.log(n);

File: 78.R
# Coin partitions

library(memoise)

Args    <- commandArgs()
divisor <- ifelse(is.na(Args[6]), 1e6, as.numeric(Args[6]))

partitions <- memoise(function(n) {
    if ( n == 0 ) {
        return(1)
    }

    parts <- 0
    k <- 1
    while ( TRUE ) {
        pent <- pentagonal(k)
        if ( n - pent < 0 ) {
            break
        }
        parts <- parts + (-1) ^ (k - 1) * partitions(n - pent)
        parts <- parts %% divisor
        if (k > 0) {
            k <- -k
        } else {
            k <- -k + 1
        }
    }

    return(parts)
})

pentagonal <- memoise(function(n) {
    return((3 * n * n - n) / 2)
})

n <- 1
while ( partitions(n) ) {
    n <- n + 1
}

cat(n, fill=TRUE)
